### PR TITLE
Adding Oauth Response for access tokens

### PR DESCRIPTION
**Pull request by almalee24**
Add new OAuth Response error handling to PayTrace, Quickbooks, Simetrik, Alelo, CheckoutV2 and Airwallex.

### CODE REVIEW

File: lib/active_merchant/billing/gateways/checkout_v2.rb [hunk 1]
```diff
@@ -341,8 +341,14 @@ def access_token_url
...
         request = 'grant_type=client_credentials'
-        response = parse(ssl_post(access_token_url, request, access_token_header))
-        response['access_token']
+        begin
+          raw_response = ssl_post(access_token_url, request, access_token_header)
+        rescue ResponseError => e
+          raise OAuthResponseError.new(e)
+        else
+          response = parse(raw_response)
+          response['access_token']
```
Comment (aenand) [inline lib/active_merchant/billing/gateways/checkout_v2.rb:350]: Should we raise the `OAuthResponseError` if `response['access_token']` is not present?

File: lib/active_merchant/billing/gateways/alelo.rb [hunk 2]
```diff
@@ -110,8 +110,14 @@ def fetch_access_token
...
 
-        parsed = parse(ssl_post(url('captura-oauth-provider/oauth/token'), post_data(params), headers))
-        Response.new(true, parsed[:access_token], parsed)
+        begin
+          raw_response = ssl_post(url('captura-oauth-provider/oauth/token'), post_data(params), headers)
+        rescue ResponseError => e
+          raise OAuthResponseError.new(e)
+        else
+          parsed = parse(raw_response)
+          Response.new(true, parsed[:access_token], parsed)
```
Comment (aenand) [inline lib/active_merchant/billing/gateways/alelo.rb:119]: Should we check that `parsed[:access_token` is defined and if it isn't raise an `OAuthResponseError`?

File: lib/active_merchant/billing/gateways/pay_trace.rb [hunk 3]
```diff
@@ -187,10 +187,15 @@ def acquire_access_token
           'Content-Type'      => 'application/x-www-form-urlencoded'
         }
         response = ssl_post(url, data, oauth_headers)
-        json_response = JSON.parse(response)
+        json_response = parse(response)
```
Comment (aenand) [inline lib/active_merchant/billing/gateways/pay_trace.rb:190]: Why remove the `JSON.`?

**Review by aenand — CHANGES_REQUESTED** [review]
Would you be able to make some unit tests for this change?

File: lib/active_merchant/billing/gateways/checkout_v2.rb (hunk 1 above)
Comment (almalee24) [reply lib/active_merchant/billing/gateways/checkout_v2.rb:350]: Yeah that's a good idea, will change it.

File: lib/active_merchant/billing/gateways/pay_trace.rb (hunk 3 above)
Comment (almalee24) [reply lib/active_merchant/billing/gateways/pay_trace.rb:190]: Just cause parse does the same thing lol

File: lib/active_merchant/billing/gateways/airwallex.rb [hunk 4]
```diff
@@ -1,3 +1,4 @@
+require 'pry'
```
Comment (aenand) [inline lib/active_merchant/billing/gateways/airwallex.rb:1]: Pesky pry statement!

File: lib/active_merchant/billing/gateways/airwallex.rb [hunk 5]
```diff
@@ -32,7 +33,7 @@ def initialize(options = {})
         @client_id = options[:client_id]
         @client_api_key = options[:client_api_key]
         super
-        @access_token = setup_access_token
+        @access_token = options[:access_token] || setup_access_token
```
Comment (aenand) [inline lib/active_merchant/billing/gateways/airwallex.rb:36]: Why make this change?

File: lib/active_merchant/billing/gateways/checkout_v2.rb [hunk 6]
```diff
@@ -18,13 +19,13 @@ class CheckoutV2Gateway < Gateway
 
       def initialize(options = {})
         @options = options
-        @access_token = nil
+        @access_token = options[:access_token] || nil
```
Comment (aenand) [inline lib/active_merchant/billing/gateways/checkout_v2.rb:22]: When is `options[:access_token]` set?

**Review by aenand — COMMENTED** [review]
Great test coverage! I think this looks good just one question about the `options[:access_token]`. Was that done to make testing easier?

File: lib/active_merchant/billing/gateways/airwallex.rb (hunk 4 above)
Comment (almalee24) [reply lib/active_merchant/billing/gateways/airwallex.rb:1]: Good catch!

**Comment by almalee24** [comment]
> Great test coverage! I think this looks good just one question about the `options[:access_token]`. Was that done to make testing easier?

Yes, to add unit testing. Without it I can't skip the access token method for all the other units tests and it causes them to fail. It couldn't find a better way to do it, the way it was done before was calling the method within the unit test file, https://github.com/activemerchant/active_merchant/pull/4851/files#diff-f3fb17a2b3e2cdf9d21bc45cc072a7f6f740a76e2b362b2168892dc94a1904b0L3

**Pull request merged**